rootston/seat.c

```diff
@@ -544,92 +549,70 @@ void roots_seat_focus_view(struct roots_seat *seat, struct roots_view *view) {
 		return;
 	}
 
-	bool found = false;
-	struct roots_seat_view *seat_view;
-	wl_list_for_each(seat_view, &seat->views, link) {
-		if (seat_view->view == view) {
-			found = true;
-			break;
+	ssize_t view_index = -1;
+	if (view != NULL) {
+		for (size_t i = 0; i < seat->views->length; ++i) {
+			if (seat->views->items[i] == view) {
+				view_index = i;
+				break;
+			}
+		}
+		if (view_index < 0) {
+			return;
 		}
-	}
-	if (!found) {
-		return;
 	}
 
-	struct roots_seat_view *prev_focus = seat->focus;
-	seat->focus = seat_view;
+	struct roots_view *prev_focus = roots_seat_get_focused_view(seat);
+	seat->has_focus = false;
 
 	// unfocus the old view if it is not focused by some other seat
-	if (prev_focus && !input_view_has_focus(seat->input, prev_focus->view)) {
-		view_activate(prev_focus->view, false);
+	if (prev_focus != NULL && !input_view_has_focus(seat->input, prev_focus)) {
+		view_activate(prev_focus, false);
 	}
 
-	if (!seat->focus) {
+	if (view == NULL) {
 		seat->cursor->mode = ROOTS_CURSOR_PASSTHROUGH;
 		return;
 	}
 
 	view_activate(view, true);
-	wl_list_remove(&seat_view->view->link);
-	wl_list_insert(&seat->input->server->desktop->views,
-		&seat_view->view->link);
+	wl_list_remove(&view->link);
+	wl_list_insert(&seat->input->server->desktop->views, &view->link);
 
-	wl_list_remove(&seat_view->link);
-	wl_list_insert(&seat->views, &seat_view->link);
+	seat->has_focus = true;
+	wlr_list_del(seat->views, view_index);
+	wlr_list_insert(seat->views, 0, view);
 	wlr_seat_keyboard_notify_enter(seat->seat, view->wlr_surface);
 }
 
-static void seat_view_destroy(struct roots_seat_view *seat_view) {
-	struct roots_seat *seat = seat_view->seat;
-
-	if (seat->focus == seat_view) {
-		seat->focus = NULL;
-		seat->cursor->mode = ROOTS_CURSOR_PASSTHROUGH;
-	}
-
-	wl_list_remove(&seat_view->destroy.link);
-	wl_list_remove(&seat_view->link);
-	free(seat_view);
-
-	// Focus first view
-	if (!wl_list_empty(&seat->views)) {
-		struct roots_seat_view *first_seat_view = wl_container_of(
-			seat->views.next, first_seat_view, link);
-		roots_seat_focus_view(seat, first_seat_view->view);
-	}
-}
-
-static void seat_view_handle_destroy(struct wl_listener *listener, void *data) {
-	struct roots_seat_view *seat_view =
-		wl_container_of(listener, seat_view, destroy);
-	seat_view_destroy(seat_view);
-}
-
 void roots_seat_add_view(struct roots_seat *seat, struct roots_view *view) {
```

Comment:
There's no reason for this. Just add it when it is focused and remove it when it is destroyed.

Reply:
We may want to give some views to some seats only.